syntax = "proto3";

package nighthawk.client;

import "google/protobuf/duration.proto";
import "google/protobuf/wrappers.proto";
import "envoy/config/core/v3alpha/base.proto";
import "envoy/extensions/transport_sockets/tls/v3alpha/cert.proto";
import "validate/validate.proto";

// Allows for static configuration of requests that should be send by the load generator.
message RequestOptions {
  envoy.config.core.v3alpha.RequestMethod request_method = 1;
  repeated envoy.config.core.v3alpha.HeaderValueOption request_headers = 2;
  // Our StreamDecoder depends on bounding the size here, so if this changes, an amendment
  // to that is needed as well.
  google.protobuf.UInt32Value request_body_size = 3 [(validate.rules).uint32 = {lte: 4194304}];
}

// Configures a remote gRPC source that will deliver to-be-replayed request data to Nighthawks
// workers.
message RequestSource {
  string uri = 19; // [(validate.rules).string.uri = true];
}

// We wrap all values so we can have a unified way of option handling with respect to
// defaults, merging, etc. As there's no stock concept for enumerations, we manually
// define custom wrappers for them. These used to be strings, which did provide the
// wrapped type.
message AddressFamily {
  enum AddressFamilyOptions {
    AUTO = 0;
    V4 = 1;
    V6 = 2;
  }
  AddressFamilyOptions value = 1;
}

message Verbosity {
  enum VerbosityOptions {
    DEFAULT = 0;
    INFO = 1;
    TRACE = 2;
    DEBUG = 3;
    WARN = 4;
    ERROR = 5;
    CRITICAL = 6;
  }
  VerbosityOptions value = 1;
}

message OutputFormat {
  enum OutputFormatOptions {
    DEFAULT = 0;
    JSON = 1;
    HUMAN = 2;
    YAML = 3;
    DOTTED = 4;
    FORTIO = 5;
  }
  OutputFormatOptions value = 1;
}

message SequencerIdleStrategy {
  enum SequencerIdleStrategyOptions {
    DEFAULT = 0;
    SPIN = 1;
    POLL = 2;
    SLEEP = 3;
  }
  SequencerIdleStrategyOptions value = 1;
}

message MultiTarget {
  message Endpoint {
    google.protobuf.StringValue address = 1;
    google.protobuf.UInt32Value port = 2 [(validate.rules).uint32 = {gte: 1, lte: 65535}];
  }
  // Whether to use HTTPS in requests to all backends; otherwise HTTP.
  google.protobuf.BoolValue use_https = 1;
  // One or more address-port pairs to receive traffic distributed with round robin.
  repeated Endpoint endpoints = 2;
  // The absolute HTTP request path (the part of the URL after host:port, e.g. /x/y/z).
  // A single path is requested from all backends. Ignored when using a RequestSource.
  google.protobuf.StringValue path = 3;
}

message H1ConnectionReuseStrategy {
  enum H1ConnectionReuseStrategyOptions {
    DEFAULT = 0;
    MRU = 1;
    LRU = 2;
  }
  H1ConnectionReuseStrategyOptions value = 1;
}

// TODO(oschaaf): Ultimately this will be a load test specification. The fact that it
// can arrive via CLI is just a concrete detail. Change this to reflect that.
message CommandLineOptions {
  // See :option:`--rps` for details.
  google.protobuf.UInt32Value requests_per_second = 1
      [(validate.rules).uint32 = {gte: 1, lte: 1000000}];
  // See :option:`--connections` for details.
  google.protobuf.UInt32Value connections = 2 [(validate.rules).uint32 = {gte: 1, lte: 1000000}];
  // See :option:`--duration` for details.
  google.protobuf.Duration duration = 3 [(validate.rules).duration.gte.nanos = 1];
  // See :option:`--timeout` for details.
  google.protobuf.Duration timeout = 4 [(validate.rules).duration.gte.seconds = 0];
  // See :option:`--h2` for details.
  google.protobuf.BoolValue h2 = 5;
  // See :option:`--concurrency` for details.
  google.protobuf.StringValue concurrency =
      6; // [(validate.rules).string = {pattern: "^([0-9]*|auto)$"}];
  // See :option:`--verbosity` for details.
  Verbosity verbosity = 7;
  // See :option:`--output-format` for details.
  OutputFormat output_format = 8;
  // See :option:`--prefetch-connections` for details.
  google.protobuf.BoolValue prefetch_connections = 9;
  // See :option:`--burst-size` for details.
  google.protobuf.UInt32Value burst_size = 10 [(validate.rules).uint32 = {lte: 1000000}];

  // See :option:`--address-family` for details.
  AddressFamily address_family = 11;
  // Either requests will be statically configured, or delivered through a remote gRPC service.
  oneof oneof_request_options {
    // See :option:`--request_options` for details.
    RequestOptions request_options = 12;
    // See :option:`--request_source` for details.
    RequestSource request_source = 26;
  }
  // See :option:`--tls_context` for details.
  envoy.extensions.transport_sockets.tls.v3alpha.UpstreamTlsContext tls_context = 13;
  // See :option:`--max-pending_requests` for details.
  google.protobuf.UInt32Value max_pending_requests = 14;
  // See :option:`--max-active_requests` for details.
  google.protobuf.UInt32Value max_active_requests = 15 [(validate.rules).uint32 = {gte: 1}];
  // See :option:`--max-requests-per-connection` for details.
  google.protobuf.UInt32Value max_requests_per_connection = 16 [(validate.rules).uint32 = {gte: 1}];
  // See :option:`--sequencer-idle-strategy` for details.
  SequencerIdleStrategy sequencer_idle_strategy = 17;
  // Either a single URI is configured, or the same traffic can be spread across a static
  // set of backends.
  oneof oneof_uri {
    // See :option:`--uri` for details.
    google.protobuf.StringValue uri = 18; // [(validate.rules).string.uri = true];
    // For details see:
    // - :option:`--multi-target-endpoint`
    // - :option:`--multi-target-path`
    // - :option:`--multi-target-use-https`
    MultiTarget multi_target = 29;
  }
  // See :option:`--trace` for details.
  google.protobuf.StringValue trace = 19; // [(validate.rules).string.uri = true];
  // See :option:`--experimental-h1-connection-reuse-strategy` for details.
  H1ConnectionReuseStrategy experimental_h1_connection_reuse_strategy = 23;
  // See :option:`--termination-predicates` for details.
  map<string, uint64> termination_predicates = 20;
  // See :option:`--failure-predicates` for details.
  map<string, uint64> failure_predicates = 24;
  // See :option:`--open-loop` for details.
  google.protobuf.BoolValue open_loop = 21;
  // See :option:`--jitter-uniform`
  google.protobuf.Duration jitter_uniform = 25 [(validate.rules).duration.gte.nanos = 1];
  // See :option --nighthawk-service
  google.protobuf.StringValue nighthawk_service = 31; // [(validate.rules).string.uri = true];
  // See :option:`--experimental-h2-use-multiple-connections` for details.
  google.protobuf.BoolValue experimental_h2_use_multiple_connections = 30;
  // See :option:`--labels`
  repeated string labels = 28;
  // See :option:`--transport-socket` for details.
  envoy.config.core.v3alpha.TransportSocket transport_socket = 27;
}
